compiler/rustc_mir_build/src/build/expr/as_place.rs

```diff
@@ -83,20 +84,25 @@ fn convert_to_hir_projections_and_truncate_for_capture<'tcx>(
     mir_projections: &[PlaceElem<'tcx>],
 ) -> Vec<HirProjectionKind> {
     let mut hir_projections = Vec::new();
+    let mut variant = None;
 
     for mir_projection in mir_projections {
         let hir_projection = match mir_projection {
             ProjectionElem::Deref => HirProjectionKind::Deref,
             ProjectionElem::Field(field, _) => {
                 // We will never encouter this for multivariant enums,
                 // read the comment for `Downcast`.
-                HirProjectionKind::Field(field.index() as u32, VariantIdx::new(0))
+                let variant = variant.unwrap_or(VariantIdx::new(0));
+                HirProjectionKind::Field(field.index() as u32, variant)
             }
-            ProjectionElem::Downcast(..) => {
-                // This projections exist only for enums that have
-                // multiple variants. Since such enums that are captured
-                // completely, we can stop here.
-                break;
+            ProjectionElem::Downcast(.., idx) => {
+                // This projections exist for enums that have
+                // single and multiple variants.
+                // For single variants, enums are not captured completely.
+                // We keep track of VariantIdx so we can use this information
+                // if the next ProjectionElem is a Field.
+                variant = Some(*idx);
```

Comment:
If we capture a multi-variant enum, we will do it in entierty in typchk, i.e. they will be trunacted by the time we get here

Reply:
Though i think the comment should be something like, "We capture multi-varirant enums completely, but we might see a downcast projection in case of single variant enums, so we need to account for it here."

We just happen to be accounting for it more generally. Since we search for ancestors in the list of captures of the place returned by this function, I think its fine to do what we are doing here.

Reply:
I think the "we capture multi-variant enums completely" is actually sort of confusingly phrased. I would say something like:

> We don't expect to see multi-variant enums here, as earlier phases will have truncated them already. However, there can still be downcasts, thanks to single-variant enums.

If this is correct, can we add an assertion that `idx` is 0?

Reply:
I dont think that is quite right. We are going from a MIR place -> HIR place. For all the HIR places we have in our capture set, we would stop at the multivairant enum and not capture anything on top of it. However the MIR place might be for something precise, eg: 

```rust
struct Foo(String, String)
x: Option<Foo>
let c = || match x {
Some(Foo(, y)) => // Upvar(x), [downcast(0), field(0), field(1)]
// ...
```

The ideal thing to do here would really be look at the type and see if it's a multi-variant enum and break, but that is a little hard to do here -- because we don't know the actual place since we don't know the capture index, and hence can't get the type. We can make it work by checking if we know any anscetors of the place as it exists right now. This sort of makes the code hard to follow and somewhat cyclic in logic where to find a valid ancestor we need to convert the mir place into some hir place and to do that we need to find a valid ancestor.  

On the otherhand, the field projection in HirPlace is defined as `Field(field_idx, variant_idx)` which as I understand is Downcast to `variant_idx` and then read the field with index `field_idx`. Which is cleaner to express and somewhat more precise. We can then make the function that compares two places to check for the ancestor relation to handle this case properly, and since our capture rules don't really capture multivariant enums it doesn't have to do much for it.

Another thing to note here is that a multivariant enum has a variant with index 0, so checking for 0 is not an "if and only if" like check.

Reply:
If we agree on this approach to thinking about the problem, we would need to update the comments to reflect this